src/lib/RpcApi.ts

```diff
@@ -98,17 +103,29 @@ export default class RpcApi {
             undefined, // preserveRequests: keep default behavior, which is true for redirects but false for postMessage
             handleHistoryBack,
         );
+        if (!handleHistoryBack) {
+            // The Keyguard client rejects on history back only if handleHistoryBack is activated. If the Keyguard does
+            // not reject it also does not provide us the localState to recover. For this case, we encode it manually in
+            // the history, to retrieve it from there.
+            this._exportStateToHistory();
+        }
```

Comment:
This assumes that the Keyguard client is always created just-in-time when it's invoked. This is currently the case, but not enforced. So this might export an incomplete state to the history, when changes to the state are done after the Keyguard client is created via this function.

Reply:
Furthermore, this now stores the very same data as the KeyguardClient stores anyway (in localstorage I believe) in a second location. Is that necessary? Can we get rid of the duplication? Does anyone remember why the RPC client is not storing its `localState` in the history object but in localstorage instead?

Reply:
The history state is exported exactly at the time the Keyguard Client is created with the rpc state.
To achieve this consistency was the idea of this change.

The data is duplicated yes. But it is only stored in the history or sessionStorage, i.e. no persistent data is duplicated. That's ok in my opinion.
Also the history state and Keyguard rpc state become relevant in different situations:
- The keyguard state is relevant when returning from the Keyguard after a successful or failed request.
- The history state is relevant for enabling reloading of pages within the Hub without loosing the exported state. This is relevant for several requests, for example Ledger requests that don't go through the Keyguard but also other requests that make use of the static store like requests that store an originalRoute.
- The history state is also used to restore the state when coming back from the Keyguard by a simple back navigation without that the Keyguard resolves or rejects the Keyguard request in which case the Keyguard does not provide the state. The goal is thus to be able to restore the state regardless of whether `handleHistoryBack` is set or not. This third bullet point is what the lines that this comment has been attached to achieve.

The Keyguard / rpc lib stores the state in the sessionStorage as it needs to be able to restore the data after a Keyguard requests finished, which will be another history entry than the one that invoked the Keyguard.

Reply:
The idea of exporting to history state grew on me in the last days. I'm accepting it.